### PR TITLE
polish the method LoadRules() in system rule manager.go (#71)

### PR DESCRIPTION
**Pull request by hellosmallstone**
Remove the channel ruleChan which seems to have no obvious meaning.
LoadRules() call onRuleUpdate() to load system rules.
Add ClearRules() to clear all the previous rules.

Signed-off-by: xu wu <wuxu1103@163.com>

### Describe what this PR does / why we need it

The implementation of LoadRules() in rule_manager.go might be not elegant, and the channel ruleChan seems to have no obvious meaning.

### Does this pull request fix one issue?

Fixes 71

### Describe how you did it

 Remove the channel ruleChan which seems to have no obvious meaning.
 LoadRules() call onRuleUpdate() to load system rules.
 Add ClearRules() to clear all the previous rules.

### Describe how to verify it

go test runs ok.

### Special notes for reviews
logRuleUpdate() seems to have no obvious meaning, therefore I delete it and add some logs in its caller.

### CODE REVIEW

File: core/system/rule_manager.go [hunk 1]
```diff
@@ -55,9 +31,28 @@ func GetRules() []*SystemRule {
 }
 
 // LoadRules loads given system rules to the rule manager, while all previous rules will be replaced.
-func LoadRules(rules []*SystemRule) (bool, error) {
-	ruleChan <- rules
-	return true, nil
+func LoadRules(rules []*SystemRule) bool {
```
Comment (louyuting) [inline core/system/rule_manager.go]: It's better to pass up the error to the upper layer.

File: core/system/rule_manager.go [hunk 2]
```diff
@@ -55,9 +31,28 @@ func GetRules() []*SystemRule {
...
+	if err := onRuleUpdate(rules); err != nil {
+		logger.Errorf("load rules %v failed: %s", rules, err.Error())
+		return false
+	}
+
+	return true
+}
+
+// ClearRules clear all the previous rules
+func ClearRules() bool {
```
Comment (louyuting) [inline core/system/rule_manager.go]: Could you please provide the use case for function "ClearRules"?

File: core/system/rule_manager.go (hunk 1 above)
Comment (hellosmallstone) [reply core/system/rule_manager.go]: In the current case, the meaning of the two return values is repeated. Ehen bool is true, error is always nil. Therefore I delete the error.
If we need pass up the error to the upper layer, we can delete the bool. What do you think of this way?

File: core/system/rule_manager.go [hunk 3]
```diff
@@ -67,22 +62,13 @@ func onRuleUpdate(rules []*SystemRule) error {
 	defer ruleMapMux.Unlock()
 
 	ruleMap = m
-	logRuleUpdate(m)
-
-	return nil
-}
-
-func logRuleUpdate(m RuleMap) {
```
[the remark this answers is not in the document]
Comment (hellosmallstone) [reply core/system/rule_manager.go]: The function buildRuleMap ignore the invalid rules and get all the valid rules. I think if the input parameter rules has invalid rules, which means that the parameters is problematic, the function should discard all the rules and return error. 
We shouldn't just discard the invalid rule and go against the original will of the caller, even though it is a wrong will.

File: core/system/rule_manager.go [hunk 4]
```diff
@@ -55,9 +31,28 @@ func GetRules() []*SystemRule {
...
+		logger.Errorf("load rules %v failed: %s", rules, err.Error())
+		return false
+	}
+
+	return true
+}
+
+// ClearRules clear all the previous rules
+func ClearRules() bool {
+	if err := onRuleUpdate(nil); err != nil {
```
[the remark this answers is not in the document]
Comment (hellosmallstone) [reply core/system/rule_manager.go]: I think the operation of ruleMap should be centralized, that is, modify it with one function, and try not to modify it with multiple functions. This seems to have some benefits for others to read the code.
    `func onRuleUpdate(r RuleMap) error` 
We can change the input parameter's type from []*SystemRule to RuleMap, and the funtion will no longer call function buildRuleMap. In this way, the meaning of the function will be more clear.

File: core/system/rule_manager.go (hunk 2 above)
Comment (hellosmallstone) [reply core/system/rule_manager.go]: In my design, the LoadRules function will discard empty rules, but the original design can set ruleMap to be empty. Therefore, I provided the ClearRules function to do it specially.
For the use case of the function, I haven't thought much about it yet. We can have some discussion about this.

File: core/system/rule_manager.go [hunk 5]
```diff
@@ -55,9 +31,28 @@ func GetRules() []*SystemRule {
...
-	ruleChan <- rules
-	return true, nil
+func LoadRules(rules []*SystemRule) bool {
+	if len(rules) == 0 {
+		logger.Debug("the rules to be loaded is empty")
+		return true
+	}
+
+	if err := onRuleUpdate(rules); err != nil {
+		logger.Errorf("load rules %v failed: %s", rules, err.Error())
```
[the remark this answers is not in the document]
Comment (hellosmallstone) [reply core/system/rule_manager.go]: Ok, I will fix it.

File: core/system/rule_manager.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (hellosmallstone) [reply core/system/rule_manager.go]: @louyuting Hi, what do you think about my thoughts?

File: core/system/rule_manager.go (hunk 1 above)
Comment (sczyh30) [reply core/system/rule_manager.go]: Actually, the `bool` here was designed to indicate whether the internal map has been changed, not for the error.

File: core/system/rule_manager.go [hunk 6]
```diff
@@ -55,9 +31,28 @@ func GetRules() []*SystemRule {
 }
 
 // LoadRules loads given system rules to the rule manager, while all previous rules will be replaced.
-func LoadRules(rules []*SystemRule) (bool, error) {
-	ruleChan <- rules
-	return true, nil
+func LoadRules(rules []*SystemRule) bool {
+	if len(rules) == 0 {
+		logger.Debug("the rules to be loaded is empty")
+		return true
```
Comment (sczyh30) [inline core/system/rule_manager.go]: It's good to have a `ClearRules` method, but we need to support clear rules by passing an empty slice to `LoadRules` method.

Comment (louyuting) [reply core/system/rule_manager.go]: It seems that the problem had been fixed.

**Comment by hellosmallstone** [comment]
@louyuting For your question, I added some descriptions.
If one caller want to update four rules, he writes one of them wrong for some reason. Updating only three rules does not seem to be the intention of the caller.
When updating only one rule, if the rule is wrong, we will clear the original rule. This operation seems to require more discussion.

**Comment by hellosmallstone** [comment]
@sczyh30 TBR, thx.

**Comment by sczyh30** [comment]
Could you please resolve the conflicts?

**Comment by hellosmallstone** [comment]
> Could you please resolve the conflicts?

Yeah, I have fixed the conflicts.

**Comment by sczyh30** [comment]
Thanks for contributing!

**Pull request merged**